### ISSUE TITLE
Generator output has only two face types

### ISSUE THREAD

**Issue by ProgramItUp** [carpedm20/BEGAN-tensorflow#30]
Anyone else having trouble training on the default CelebA dataset?

After training on the CelebA dataset for 450,000 iterations, or about 33 hours, with the default settings, I had a look at the output and noticed something odd.    The generated images, *_G.png, had only two face types in each file.  Between files they were different, but there were always just two distinct faces.   

The output from the model tests is the same, just two faces:
`pytthon main.py --dataset=CelebA --load_path=CelebA_0625_080047 --use_gpu=True --is_train=False --split valid
`
Is this an example of mode collapse?  
Can anyone suggest better training parameters?

**Comment by Tony-Chiong** [comment]
I recommend you to adopt fwiffo's advice:  change --d_lr to 0.00004, change --g_lr to 0.00004, reduce --lr_update_step, (you can also set --batch_size to 4 for faster training.)

see: [https://github.com/carpedm20/BEGAN-tensorflow/issues/1](url)

**Comment by ProgramItUp** [comment]
@Tony-Chiong : Thank you for the clear advice, it really helped.  The CelebA dataset is working much better!  

For a different dataset  with only a few thousand images I set d_lr and g_lr = 0.00004, and even as low as  0.00001.  The first 20k to 50k iterations all of the output image types real, fake and generated progress nicely and show steady improvement.  After about 100k iterations the fake and generated images look much worse than at 50k iterations and and have gone into mode collapse.   The real images continue to improve and capture details in the original data.

Does it sound like the learning rate for the generator not changing fast enough?
Thoughts welcomed.

**Comment by civilman628** [comment]
I meet the same problem use default config and train a set of bag images. I can only see 2 types bags in G.png. there are not diversified.

**Comment by henrych4** [comment]
Similar problem when trained with CelebA dataset. I have trained it for 500k iterations and before 410k iterations, everything looks fine but soon mode collapse problem occurs. Only three types of face images are shown in a batch (16) of images.

**Comment by duduheihei** [comment]
When I look through the source code, I notice that "z" , the input of "Generator", doesn't change during training process. Maybe this results mode collapse.  You can check it.

**Comment by zacharynevin** [comment]
@duduheihei The `z` input does change. `tf.random_normal` changes every time `sess.run` is called (https://github.com/carpedm20/BEGAN-tensorflow/blob/master/trainer.py#L143).
